src/keypad.py

```diff
@@ -0,0 +1,488 @@
+        If there are no queued events, do not touch ``event`` and return ``False``.
+
+        The advantage of this method over ``get()`` is that it does not allocate storage.
+        Instead you can reuse an existing ``Event`` object.
+
+        Note that the queue size is limited; see ``max_events`` in the constructor of
+        a scanner such as `Keys` or `KeyMatrix`.
+
+        :return ``True`` if an event was available and stored, ``False`` if not.
+        :rtype: bool
+        """
+        if not self._events:
+            return False
+        next_event = self._events.popleft()
+        # pylint: disable=protected-access
+        event._key_number = next_event._key_number
+        event._pressed = next_event._pressed
+        # pylint: enable=protected-access
+        return True
+
+    def clear(self):
+        """
+        Clear any queued key transition events. Also sets `overflowed` to ``False``.
+        """
+        self._events.clear()
+        self._overflowed = False
+
+    def __bool__(self):
+        """``True`` if `len()` is greater than zero.
+        This is an easy way to check if the queue is empty.
+        """
+        return len(self._events) > 0
+
+    def __len__(self):
+        """Return the number of events currently in the queue. Used to implement ``len()``."""
+        return len(self._events)
+
+    @property
+    def overflowed(self):
+        """
+        ``True`` if an event could not be added to the event queue because it was full. (read-only)
+        Set to ``False`` by  `clear()`.
+        """
+        return self._overflowed
+
+    def keypad_eventqueue_record(self, key_number, current):
+        """Record a new event"""
+        if len(self._events) == self._events.maxlen:
+            self._overflowed = True
+        else:
+            self._events.append(Event(key_number, current))
+
+
+class _KeysBase:
+    def __init__(self, interval, max_events, scanning_function):
+        self._interval = interval
+        self._last_scan = time.monotonic()
+        self._events = _EventQueue(max_events)
+        self._scanning_function = scanning_function
+        self._scan_thread = threading.Thread(target=self._scanning_loop, daemon=True)
+        self._scan_thread.start()
+
+    @property
+    def events(self):
+        """The EventQueue associated with this Keys object. (read-only)"""
+        return self._events
+
+    def deinit(self):
+        """Stop scanning"""
+        if self._scan_thread.is_alive():
+            self._scan_thread.join()
+
+    def __enter__(self):
+        """No-op used by Context Managers."""
+        return self
+
+    def __exit__(self, exception_type, exception_value, traceback):
+        """
+        Automatically deinitializes when exiting a context. See
+        :ref:`lifetime-and-contextmanagers` for more info.
+        """
+        self.deinit()
+
+    def _scanning_loop(self):
+        while True:
+            self._scanning_function()
+            time.sleep(0.001)
```

Comment:
Suppose you just call the scanning function once per interval here? Then you don't need to check in each internal scanning function. In the C implementation, this was not as easy, but it's easier here.
```suggestion
            time.sleep(self._interval)
```

Reply:
Though this simplifies things, it loses accuracy because it is sleep interval + processing time. I could probably do something more accurate like grabbing time.sleep(self._interval - (time.monotonic - self._last_scan)).

Reply:
I was thinking this would use more CPU (polling the scanning routines every msec instead of, say, every 20msecs). We are talking about msec or 10-msec granularity, so I'm not too worried about accuracy. But it's not a big deal -- I'm happy to  approve as is.

Reply:
True that it's just a tiny sliver of time, but I do like to make it as accurate as possible :) Anyway, I updated in the spirit of your suggestion.